pkg/controller/servicednsendpoint/servicednsendpoint_controller.go

```diff
@@ -0,0 +1,464 @@
+		targets := [][]string{zoneTargets, regionTargets, globalTargets}
+
+		for i, target := range targets {
+			endpoint, err := generateEndpoint(dnsNames[i], target, dnsNames[i+1])
+			if err != nil {
+				return nil, err
+			}
+			endpoints = append(endpoints, endpoint)
+		}
+	}
+
+	return endpoints, nil
+}
+
+// getHealthyTargets returns the hostnames and/or IP addresses of healthy endpoints for the service, at a zone, region and global level (or an error)
+func getHealthyTargets(zone, region string, dnsObject *feddnsv1a1.MultiClusterServiceDNSRecord) (zoneTargets, regionTargets, globalTargets feddnsv1a1.Targets) {
+	// If federated dnsObject is deleted, return empty endpoints, so that DNS records are removed
+	if dnsObject.DeletionTimestamp != nil {
+		return zoneTargets, regionTargets, globalTargets
+	}
+
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		if clusterDNS.Zone == zone {
+			zoneTargets = append(zoneTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+		}
+	}
+
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		if clusterDNS.Region == region {
+			regionTargets = append(regionTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+		}
+	}
+
+	for _, clusterDNS := range dnsObject.Status.DNS {
+		globalTargets = append(globalTargets, ExtractLoadBalancerTargets(clusterDNS.LoadBalancer)...)
+	}
+
+	return zoneTargets, regionTargets, globalTargets
+}
+
+func generateEndpoint(name string, targets feddnsv1a1.Targets, uplevelCname string) (ep *feddnsv1a1.Endpoint, err error) {
+	ep = &feddnsv1a1.Endpoint{
+		DNSName:   name,
+		RecordTTL: minDNSTTL,
+	}
+
+	if len(targets) > 0 {
+		targets, err = getResolvedTargets(targets, netWrapper)
+		if err != nil {
+			return nil, err
+		}
+		ep.Targets = targets
+		ep.RecordType = RecordTypeA
+	} else {
+		ep.Targets = []string{uplevelCname}
+		ep.RecordType = RecordTypeCNAME
+	}
+
+	return ep, nil
+}
+
+// getResolvedTargets performs DNS resolution on the provided slice of endpoints (which might be DNS names
+// or IPv4 addresses) and returns a list of IPv4 addresses.  If any of the endpoints are neither valid IPv4
+// addresses nor resolvable DNS names, non-nil error is also returned (possibly along with a partially
+// complete list of resolved endpoints.
+func getResolvedTargets(targets feddnsv1a1.Targets, netWrapper NetWrapper) (feddnsv1a1.Targets, error) {
+	resolvedTargets := sets.String{}
+	for _, target := range targets {
+		if net.ParseIP(target) == nil {
+			// It's not a valid IP address, so assume it's a DNS name, and try to resolve it,
+			// replacing its DNS name with its IP addresses in expandedEndpoints
+			// through an interface abstracting the internet
+			ipAddrs, err := netWrapper.LookupHost(target)
+			if err != nil {
+				glog.Errorf("Failed to resolve %s, err: %v", target, err)
+				return resolvedTargets.List(), err
+			}
+			for _, ip := range ipAddrs {
+				resolvedTargets = resolvedTargets.Union(sets.NewString(ip))
```

Comment:
Assuming you're intending to insert this new item into the set, is there a reason to avoid `Insert`?

```golang
resolvedTargets.Insert(ip)
```

Reply:
I was under the impression that Insert would just insert even if it is a duplicate. Updated now to use Insert.

Reply:
A property of a set is that it does not contain duplicates, so adding a value that already exists is effectively a no-op.